File: sdks/node/src/local.ts
/**
 * LocalClient — poll hub-router for requests, process them, push results back.
 * Fully async, non-blocking. The poll loop runs as an independent Promise so it
 * never blocks your application's event loop.
 */
import { HubRouterError, RequestNotFoundError } from "./errors.js";
import type {
  LocalClientOptions,
  ProcessorFn,
  PullResponse,
  QueuedRequest,
  Result,
} from "./types.js";

const DEFAULT_BATCH_SIZE = 10;
const DEFAULT_POLL_INTERVAL_MS = 1_000;
const DEFAULT_WORKERS = 1;

export class LocalClient {
  private readonly baseUrl: string;
  private readonly apiKey: string;
  private readonly batchSize: number;
  private readonly pollIntervalMs: number;
  private readonly workers: number;
  private stopped = false;

  /**
   * @param baseUrl   hub-router base URL, e.g. "https://hub.example.com"
   * @param apiKey    X-Local-API-Key value
   * @param options   Optional overrides
   */
  constructor(
    baseUrl: string,
    apiKey: string,
    options: LocalClientOptions = {}
  ) {
    this.baseUrl = baseUrl.replace(/\/$/, "");
    this.apiKey = apiKey;
    this.batchSize = options.batchSize ?? DEFAULT_BATCH_SIZE;
    this.pollIntervalMs = options.pollInterval ?? DEFAULT_POLL_INTERVAL_MS;
    this.workers = options.workers ?? DEFAULT_WORKERS;
  }

  /**
   * Start the poll loop. Returns a Promise that resolves when {@link stop} is called.
   *
   * **Non-blocking usage** — run alongside your server without await:
   * ```ts
   * client.run(processor);          // fire and forget
   * await server.listen(3000);      // your server keeps running
   * ```
   *
   * **Blocking usage** — run as the sole task:
   * ```ts
   * await client.run(processor);
   * ```
   */
  async run(processor: ProcessorFn): Promise<void> {
    // Simple semaphore: track active worker count.
    let activeWorkers = 0;
    const waitForWorker = (): Promise<void> =>
      new Promise((resolve) => {
        const check = () => {
          if (activeWorkers < this.workers) {
            resolve();
          } else {
            setImmediate(check);
          }
        };
        check();
      });

    while (!this.stopped) {
      let requests: QueuedRequest[];
      try {
        requests = await this.pullBatch();
      } catch {
        // Transient error — wait and retry.
        await this._sleep(this.pollIntervalMs);
        continue;
      }

      if (requests.length === 0) {
        await this._sleep(this.pollIntervalMs);
        continue;
      }

      for (const req of requests) {
        await waitForWorker();
        if (this.stopped) break;
        activeWorkers++;
        // Dispatch without await — run concurrently, bounded by semaphore.
        this._processAndPush(processor, req).finally(() => {
          activeWorkers--;
        });
      }
    }

    // Drain: wait for all active workers to finish.
    while (activeWorkers > 0) {
      await this._sleep(50);
    }
  }

  /** Signal the poll loop to stop after the current batch finishes. */
  stop(): void {
    this.stopped = true;
  }

  /** Fetch up to `batchSize` pending requests. Never blocks the event loop. */
  async pullBatch(): Promise<QueuedRequest[]> {
    const res = await fetch(
      `${this.baseUrl}/queue/pull?batch=${this.batchSize}`,
      { headers: { "X-Local-API-Key": this.apiKey } }
    );

    if (res.status !== 200) {
      throw new HubRouterError(
        `unexpected status ${res.status}: ${await res.text()}`
      );
    }

    const data = (await res.json()) as PullResponse;
    return data.requests ?? [];
  }

  /** Push a completed result to hub-router. */
  async pushResult(result: Result): Promise<void> {
    const body: Result = {
      ...result,
      completed_at: result.completed_at ?? new Date().toISOString(),
    };

    const res = await fetch(`${this.baseUrl}/queue/result`, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        "X-Local-API-Key": this.apiKey,
      },
      body: JSON.stringify(body),
    });

    if (res.status === 404) {
      throw new RequestNotFoundError(result.request_id);
    }
    if (res.status !== 204) {
      throw new HubRouterError(
        `unexpected status ${res.status}: ${await res.text()}`
      );
    }
  }

  private async _processAndPush(
    processor: ProcessorFn,
    req: QueuedRequest
  ): Promise<void> {
    let result: Result;
    try {
      result = await processor(req);
    } catch (err) {
      const message = err instanceof Error ? err.message : String(err);
      result = {
        request_id: req.id,
        payload: { error: message },
        status_code: 500,
        error: message,
      };
    }
    try {
      await this.pushResult(result);
    } catch {
      // Best-effort: request may have expired.
    }
  }

  private _sleep(ms: number): Promise<void> {
    return new Promise((resolve) => setTimeout(resolve, ms));
  }
}

File: pkg/client/local.go
package client

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"strings"
	"sync"
	"time"

	"github.com/jaar23/hub-router/internal/model"
)

// ProcessorFunc is the callback the local server provides to handle a request.
// It receives a QueuedRequest and must return a Result (or an error).
// Returning an error causes a Result with StatusCode=500 and the error message to be pushed.
type ProcessorFunc func(ctx context.Context, req *model.QueuedRequest) (*model.Result, error)

// LocalClient polls hub-router for pending requests, processes them, and pushes results back.
// It is safe for concurrent use.
type LocalClient struct {
	baseURL string
	apiKey  string
	opts    LocalOptions
}

// NewLocalClient creates a LocalClient.
// baseURL should be the hub-router base URL. apiKey is the X-Local-API-Key value.
func NewLocalClient(baseURL, apiKey string, opts ...func(*LocalOptions)) *LocalClient {
	o := defaultLocalOptions()
	for _, fn := range opts {
		fn(&o)
	}
	return &LocalClient{
		baseURL: strings.TrimRight(baseURL, "/"),
		apiKey:  apiKey,
		opts:    o,
	}
}

// WithBatchSize sets the number of requests to pull per cycle.
func WithBatchSize(n int) func(*LocalOptions) {
	return func(o *LocalOptions) { o.BatchSize = n }
}

// WithPollInterval sets the sleep duration when the queue is empty.
func WithPollInterval(d time.Duration) func(*LocalOptions) {
	return func(o *LocalOptions) { o.PollInterval = d }
}

// WithWorkers sets the number of concurrent processing goroutines.
func WithWorkers(n int) func(*LocalOptions) {
	return func(o *LocalOptions) { o.Workers = n }
}

// WithLocalHTTPClient injects a custom HTTP client.
func WithLocalHTTPClient(hc *http.Client) func(*LocalOptions) {
	return func(o *LocalOptions) { o.HTTPClient = hc }
}

// Run starts the poll loop. It blocks until ctx is cancelled.
// Each batch is dispatched concurrently, bounded by opts.Workers.
func (c *LocalClient) Run(ctx context.Context, processor ProcessorFunc) error {
	sem := make(chan struct{}, c.opts.Workers)
	var wg sync.WaitGroup

	for {
		if ctx.Err() != nil {
			break
		}

		requests, err := c.PullBatch(ctx)
		if err != nil {
			if ctx.Err() != nil {
				break
			}
			// Transient error — wait and retry.
			select {
			case <-ctx.Done():
				break
			case <-time.After(c.opts.PollInterval):
			}
			continue
		}

		if len(requests) == 0 {
			select {
			case <-ctx.Done():
				goto done
			case <-time.After(c.opts.PollInterval):
			}
			continue
		}

		for _, req := range requests {
			req := req // capture
			sem <- struct{}{}
			wg.Add(1)
			go func() {
				defer func() { <-sem; wg.Done() }()
				result := c.process(ctx, processor, req)
				// Best-effort push; ignore error on cancellation.
				_ = c.PushResult(ctx, result)
			}()
		}
	}

done:
	// Wait for in-flight processors to finish.
	wg.Wait()
	return ctx.Err()
}

// PullBatch fetches up to opts.BatchSize pending requests. Never blocks.
func (c *LocalClient) PullBatch(ctx context.Context) ([]*model.QueuedRequest, error) {
	url := fmt.Sprintf("%s/queue/pull?batch=%d", c.baseURL, c.opts.BatchSize)
	req, err := http.NewRequestWithContext(ctx, http.MethodGet, url, nil)
	if err != nil {
		return nil, err
	}
	req.Header.Set("X-Local-API-Key", c.apiKey)

	resp, err := c.opts.HTTPClient.Do(req)
	if err != nil {
		return nil, fmt.Errorf("pull batch: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		raw, _ := io.ReadAll(resp.Body)
		return nil, fmt.Errorf("unexpected status %d: %s", resp.StatusCode, string(raw))
	}

	var pullResp model.PullResponse
	if err := json.NewDecoder(resp.Body).Decode(&pullResp); err != nil {
		return nil, fmt.Errorf("decode pull response: %w", err)
	}
	return pullResp.Requests, nil
}

// PushResult submits a completed result to hub-router.
func (c *LocalClient) PushResult(ctx context.Context, result *model.Result) error {
	body, err := json.Marshal(result)
	if err != nil {
		return fmt.Errorf("marshal result: %w", err)
	}

	req, err := http.NewRequestWithContext(ctx, http.MethodPost, c.baseURL+"/queue/result", bytes.NewReader(body))
	if err != nil {
		return err
	}
	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("X-Local-API-Key", c.apiKey)

	resp, err := c.opts.HTTPClient.Do(req)
	if err != nil {
		return fmt.Errorf("push result: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode == http.StatusNotFound {
		return fmt.Errorf("request ID %s not found (may have expired)", result.RequestID)
	}
	if resp.StatusCode != http.StatusNoContent {
		raw, _ := io.ReadAll(resp.Body)
		return fmt.Errorf("unexpected status %d: %s", resp.StatusCode, string(raw))
	}
	return nil
}

// process calls the ProcessorFunc and wraps any error into a Result.
func (c *LocalClient) process(ctx context.Context, fn ProcessorFunc, req *model.QueuedRequest) *model.Result {
	result, err := fn(ctx, req)
	if err != nil {
		errBytes, _ := json.Marshal(map[string]string{"error": err.Error()})
		return &model.Result{
			RequestID:   req.ID,
			Payload:     errBytes,
			StatusCode:  500,
			Error:       err.Error(),
			CompletedAt: time.Now(),
		}
	}
	if result == nil {
		result = &model.Result{RequestID: req.ID, StatusCode: 200}
	}
	if result.CompletedAt.IsZero() {
		result.CompletedAt = time.Now()
	}
	return result
}
